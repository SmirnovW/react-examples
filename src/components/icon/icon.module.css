:local {
  .container {
    pointer-events: none;

    use {
      fill: currentColor;
      stroke: currentColor;
    }
  }
}
