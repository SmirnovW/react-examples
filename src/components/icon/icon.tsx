import React from 'react';

import * as commonIcons from './svg/common/index';
import * as fastFoodIcons from './svg/fastfood/index';
import * as tablewareIcons from './svg/tableware/index';

import style from './icon.module.css';

interface Props {
    name?: string;
    size?: string;
}

interface Icons {
    [key: string]: any;
}

interface Sizes {
    [name: string]: number | string;
}

const icons: Icons = {
    ...commonIcons,
    ...fastFoodIcons,
    ...tablewareIcons,
};

const SIZE_RESPONSIVE = 'responsive';
const SIZE_SMALL = 'small';
const SIZE_MEDIUM = 'medium';
const SIZE_LARGE = 'large';

/* Sizes values */
const SIZES: Sizes = {
    [SIZE_SMALL]: 16,
    [SIZE_MEDIUM]: 22,
    [SIZE_LARGE]: 34,
    [SIZE_RESPONSIVE]: '100%'
};

export {SIZE_LARGE, SIZE_MEDIUM, SIZE_SMALL, SIZE_RESPONSIVE};

export const Icon: React.FC<Props> = ({name = '', size = SIZE_RESPONSIVE}) => {
    const sizeValue = SIZES[size] || size;
    const icon = icons[name];

    if (typeof icon === 'undefined') {
        console.error('Requested icon does not exist');
        return null;
    }

    return (
        <svg width={sizeValue} height={sizeValue} viewBox={icon.viewBox} className={style.container}>
            <use xlinkHref={icon.url} href={icon.url} />
        </svg>
    );
};
