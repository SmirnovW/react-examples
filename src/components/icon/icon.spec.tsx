import React from 'react';
import renderer from 'react-test-renderer';

import {Icon, SIZE_SMALL} from "./icon";

jest.mock('./svg/fastfood/pizza.svg', () => ({
    url: '/path/#pizza',
    viewBox: '0 0 30 30',
}));


describe('Icon component', () => {
    it('should render the "pizza" icon', () => {
        const tree = renderer.create(<Icon name="pizza" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render the "pizza" in small size icon', () => {
        const tree = renderer.create(<Icon name="pizza" size={SIZE_SMALL} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders NULL if the `icon` does not exist', () => {
        global.console.error = jest.fn();
        const tree = renderer
            .create(<Icon name="this-name-doesn-not-exist" />)
            .toJSON();
        expect(tree).toMatchSnapshot();
        expect(global.console.error).toBeCalled();
    });
});