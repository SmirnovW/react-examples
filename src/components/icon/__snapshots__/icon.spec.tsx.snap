// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component renders NULL if the \`icon\` does not exist 1`] = `null`;

exports[`Icon component should render the "pizza" icon 1`] = `
<svg
  className="container"
  height="100%"
  viewBox="0 0 30 30"
  width="100%"
>
  <use
    href="/path/#pizza"
    xlinkHref="/path/#pizza"
  />
</svg>
`;

exports[`Icon component should render the "pizza" in small size icon 1`] = `
<svg
  className="container"
  height={16}
  viewBox="0 0 30 30"
  width={16}
>
  <use
    href="/path/#pizza"
    xlinkHref="/path/#pizza"
  />
</svg>
`;
